rust/datafusion/src/optimizer/projection_push_down.rs

```diff
@@ -0,0 +1,377 @@
+    use std::cell::RefCell;
+    use std::rc::Rc;
+    use std::sync::Arc;
+
+    #[test]
+    fn aggregate_no_group_by() {
+        let table_scan = test_table_scan();
+
+        let aggregate = Aggregate {
+            group_expr: vec![],
+            aggr_expr: vec![Column(1)],
+            schema: Arc::new(Schema::new(vec![Field::new(
+                "MAX(b)",
+                DataType::UInt32,
+                false,
+            )])),
+            input: Rc::new(table_scan),
+        };
+
+        assert_optimized_plan_eq(&aggregate, "Aggregate: groupBy=[[]], aggr=[[#0]]\n  TableScan: test projection=Some([1])");
+    }
+
+    #[test]
+    fn aggregate_group_by() {
+        let table_scan = test_table_scan();
+
+        let aggregate = Aggregate {
+            group_expr: vec![Column(2)],
+            aggr_expr: vec![Column(1)],
+            schema: Arc::new(Schema::new(vec![
+                Field::new("c", DataType::UInt32, false),
+                Field::new("MAX(b)", DataType::UInt32, false),
+            ])),
+            input: Rc::new(table_scan),
+        };
+
+        assert_optimized_plan_eq(&aggregate, "Aggregate: groupBy=[[#1]], aggr=[[#0]]\n  TableScan: test projection=Some([1, 2])");
+    }
+
+    #[test]
+    fn aggregate_no_group_by_with_selection() {
+        let table_scan = test_table_scan();
+
+        let selection = Selection {
+            expr: Column(2),
+            input: Rc::new(table_scan),
+        };
+
+        let aggregate = Aggregate {
+            group_expr: vec![],
+            aggr_expr: vec![Column(1)],
+            schema: Arc::new(Schema::new(vec![Field::new(
+                "MAX(b)",
+                DataType::UInt32,
+                false,
+            )])),
+            input: Rc::new(selection),
+        };
+
+        assert_optimized_plan_eq(&aggregate, "Aggregate: groupBy=[[]], aggr=[[#0]]\n  Selection: #1\n    TableScan: test projection=Some([1, 2])");
+    }
+
+    #[test]
+    fn cast() {
+        let table_scan = test_table_scan();
+
+        let projection = Projection {
+            expr: vec![Cast {
+                expr: Rc::new(Column(2)),
+                data_type: DataType::Float64,
+            }],
+            input: Rc::new(table_scan),
+            schema: Arc::new(Schema::new(vec![Field::new(
+                "CAST(c AS float)",
+                DataType::Float64,
+                false,
+            )])),
+        };
+
+        assert_optimized_plan_eq(
+            &projection,
+            "Projection: CAST(#0 AS Float64)\n  TableScan: test projection=Some([2])",
+        );
+    }
+
+    fn assert_optimized_plan_eq(plan: &LogicalPlan, expected: &str) {
+        let optimized_plan = optimize(plan);
+        let formatted_plan = format!("{:?}", optimized_plan);
+        assert_eq!(formatted_plan, expected);
+    }
+
+    fn optimize(plan: &LogicalPlan) -> Rc<LogicalPlan> {
+        let rule: Rc<RefCell<OptimizerRule>> =
```

Comment:
Why we need `Rc` and `RefCell` here? can we do:
```rust
        let mut rule = ProjectionPushDown::new();
        rule.optimize(plan).unwrap()
```

Reply:
You are correct. I was just trying to case from `ProjectionPushDown` to `OptimizerRule` since eventually there will be a list of rules to apply. I will simplify this for now though.